float8_experimental/float8_dynamic_linear.py

```diff
@@ -54,39 +72,38 @@ def __init__(self, **super_kwargs):
         super().__init__(**super_kwargs)
 
     def forward(self, x):
-        # cast x to float8_e4m3fn if not using activation hooks
         x_fp8 = self.cast_to_float8_e4m3fn(x)
-
-        # cast w to float8_e4m3fn
-        w_fp8 = self.cast_to_float8_e4m3fn(self.weight)
-
+        w_fp8 = (
+            self.weight
+            if isinstance(self.weight, Float8Tensor)  # cast by FSDP
+            else self.cast_to_float8_e4m3fn(self.weight)
+        )
         y = torch.nn.functional.linear(x_fp8, w_fp8, self.bias)
-
-        # Cast gradY to float8_e5m2 during backward if not using activation hooks
         y = self.cast_to_float8_e5m2_bw(y)
-
         return y
 
-    def cast_to_float8_e4m3fn(self, inpt_tensor: torch.Tensor) -> Float8Tensor:
-        if tensor_already_casted_to_fp8(inpt_tensor):
-            # check to early return if already casted to float8
-            return inpt_tensor
-        scale = tensor_to_scale(inpt_tensor, torch.float8_e4m3fn)
-        return Float8Tensor.to_float8(
-            inpt_tensor, scale, torch.float8_e4m3fn, emulate=self.emulate
-        )
+    def cast_to_float8_e4m3fn(
+        self, inpt_tensor: torch.Tensor, reduce_amax: bool = False
+    ) -> Float8Tensor:
+        return cast_to_float8_e4m3fn(inpt_tensor, self.emulate, reduce_amax)
 
     def cast_to_float8_e5m2_bw(self, gradY: torch.Tensor) -> torch.Tensor:
         return NoopFwToFloat8E5M2Bw.apply(gradY, self.emulate)
 
     @classmethod
-    def from_float(cls, mod, emulate: bool = False) -> "Float8DynamicLinear":
+    def from_float(
+        cls,
+        mod,
+        emulate: bool = False,
+        use_fsdp_fp8_all_gather: bool = False,
+    ) -> "Float8DynamicLinear":
         """
         Create an nn.Linear with fp8 compute from a regular nn.Linear
 
         Args:
             mod (torch.nn.Linear): nn.Linear to convert
             emulate (bool): whether to emulate fp8 matmul logic in float32
+            use_fsdp_fp8_all_gather (bool): whether to use fp8 all-gather for FSDP
```

Comment:
Probably more of a driss/vasiliy question, but does it make more sense for this to be a global config to avoid cluttering up the subclasses? Do we ever want to mix fp8 vs non-fp8 allgather in a single program?

Reply:
Good point! I am definitely good with making this a global flag.

(I think having the flag for some time might be helpful as we test out the fp8 all-gather in real workloads. We also may not want this subclass to affect the single-GPU / non-FSDP case.)

Reply:
I changed this to a global flag now.